File: src/arch_ops_server/firewall.py
# SPDX-License-Identifier: GPL-3.0-only OR MIT
"""
nftables firewall management module.
Manages nftables rules: listing, port management, validation, reload.
"""

import logging
from typing import Dict, Any, Optional

from .utils import (
    IS_ARCH,
    run_command,
    create_error_response,
    check_command_exists,
)

logger = logging.getLogger(__name__)

NFTABLES_CONF = "/etc/nftables.conf"


async def _nft(args: list[str]) -> tuple[int, str, str]:
    return await run_command(["nft"] + args, timeout=10, check=False)


async def manage_firewall(
    action: str,
    port: Optional[int] = None,
    proto: str = "tcp",
    interface: Optional[str] = None,
    chain: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Manage nftables firewall: list rules, add/remove ports, validate config, reload.

    Actions:
      list_rules  — nft list ruleset or specific chain
      add_port    — add a temporary port to INPUT chain
      remove_port — remove a port from INPUT chain
      validate    — check nftables.conf syntax
      reload      — reload nftables ruleset
    """
    if not IS_ARCH:
        return create_error_response(
            "NotSupported", "This feature is only available on Arch Linux."
        )

    if not check_command_exists("nft"):
        return create_error_response(
            "NotSupported",
            "nftables not installed. Install with: sudo pacman -S nftables",
        )

    if action == "list_rules":
        logger.info("Listing nftables ruleset")
        cmd = ["nft"]
        if chain:
            cmd.extend(["list", "chain", "inet", "filter", chain])
        else:
            cmd.append("list ruleset")

        exit_code, out, err = await run_command(cmd, timeout=10, check=False)
        if exit_code != 0:
            return create_error_response("CommandError", f"nft failed: {err}")

        return {
            "ruleset": out.strip(),
            "rule_count": len(
                [
                    l
                    for l in out.splitlines()
                    if l.strip() and not l.strip().startswith("#")
                ]
            ),
        }

    elif action == "add_port":
        if not port:
            return create_error_response(
                "MissingArgument", "port is required for add_port"
            )
        proto = proto.lower()
        if proto not in ("tcp", "udp"):
            return create_error_response(
                "InvalidValue", f"Invalid protocol: {proto}. Use tcp or udp."
            )

        logger.info(f"Adding port {port}/{proto} to nftables INPUT")
        rule = f"tcp dport {port}" if proto == "tcp" else f"udp dport {port}"
        if interface:
            rule = f"iif {interface} {rule}"

        exit_code, out, err = await _nft(
            [
                "add",
                "rule",
                "inet",
                "filter",
                "INPUT",
                rule,
                "accept",
            ]
        )
        if exit_code != 0:
            return create_error_response("CommandError", f"Failed to add port: {err}")

        return {
            "ok": True,
            "port": port,
            "proto": proto,
            "interface": interface,
            "message": f"Port {port}/{proto} added to INPUT chain (temporary — write to {NFTABLES_CONF} to persist)",
        }

    elif action == "remove_port":
        if not port:
            return create_error_response(
                "MissingArgument", "port is required for remove_port"
            )
        proto = proto.lower()

        logger.info(f"Removing port {port}/{proto} from nftables INPUT")

        # nft delete rule by handle — find handle first
        if proto == "tcp":
            grep_pat = f"dport {port}"
        else:
            grep_pat = f"udp dport {port}"

        exit_code, out, err = await _nft(
            ["-a", "list", "chain", "inet", "filter", "INPUT"]
        )
        if exit_code != 0:
            return create_error_response("CommandError", f"Failed to list rules: {err}")

        handle = None
        for line in out.splitlines():
            if grep_pat in line and "accept" in line:
                parts = line.strip().split()
                if "handle" in parts:
                    idx = parts.index("handle")
                    if idx + 1 < len(parts):
                        handle = parts[idx + 1]
                        break

        if not handle:
            return create_error_response(
                "NotFound",
                f"No matching rule found for port {port}/{proto} in INPUT chain",
            )

        exit_code, out, err = await _nft(
            [
                "delete",
                "rule",
                "inet",
                "filter",
                "INPUT",
                "handle",
                handle,
            ]
        )
        if exit_code != 0:
            return create_error_response(
                "CommandError", f"Failed to delete rule: {err}"
            )

        return {
            "ok": True,
            "port": port,
            "proto": proto,
            "message": f"Port {port}/{proto} removed from INPUT chain",
        }

    elif action == "validate":
        logger.info("Validating nftables configuration")
        exit_code, out, err = await run_command(
            ["nft", "-c", "-f", NFTABLES_CONF],
            timeout=5,
            check=False,
        )
        valid = exit_code == 0
        return {
            "valid": valid,
            "config_file": NFTABLES_CONF,
            "error": err.strip() if not valid else None,
            "message": "Configuration is valid"
            if valid
            else f"Validation failed: {err.strip()}",
        }

    elif action == "reload":
        logger.info("Reloading nftables")
        exit_code, out, err = await run_command(
            ["systemctl", "reload", "nftables"],
            timeout=10,
            check=False,
        )
        if exit_code != 0:
            return create_error_response(
                "CommandError", f"Failed to reload nftables: {err}"
            )

        return {"ok": True, "message": "nftables reloaded successfully"}

    else:
        return create_error_response(
            "InvalidAction",
            f"Unknown action: '{action}'. Valid actions: list_rules, add_port, remove_port, validate, reload.",
        )

File: src/tools/firewall.rs
use crate::error::ToolError;
use crate::platform::is_arch_linux;
use crate::protocol::Tool;
use async_trait::async_trait;
use serde_json::Value;
use std::collections::HashMap;

use super::ToolHandler;

pub struct ManageFirewall;

#[async_trait]
impl ToolHandler for ManageFirewall {
    fn info(&self) -> Tool {
        Tool {
            name: "manage_firewall".into(),
            description: "Manage nftables firewall: list_rules, add_port, remove_port, validate, reload".into(),
            input_schema: serde_json::json!({
                "type": "object",
                "properties": {
                    "action": { "type": "string", "enum": ["list_rules", "add_port", "remove_port", "validate", "reload"] },
                    "port": { "type": "integer", "description": "Port number" },
                    "proto": { "type": "string", "enum": ["tcp", "udp"], "default": "tcp" },
                    "interface": { "type": "string", "description": "Network interface" },
                    "chain": { "type": "string", "default": "input", "description": "nftables chain" }
                },
                "required": ["action"]
            }),
        }
    }

    async fn call(&self, args: HashMap<String, Value>) -> Result<Value, ToolError> {
        if !is_arch_linux() { return Err(ToolError::platform_not_arch()); }

        let action = args.get("action").and_then(|v| v.as_str()).ok_or_else(|| ToolError::invalid_argument("Missing: action"))?;

        match action {
            "list_rules" => {
                let result = crate::command::run("sudo", &["nft", "list", "ruleset"]).await?;
                Ok(serde_json::json!({ "ruleset": result.stdout }))
            }
            "add_port" => {
                let port = args.get("port").and_then(|v| v.as_u64()).ok_or_else(|| ToolError::invalid_argument("Missing: port"))?;
                let proto = args.get("proto").and_then(|v| v.as_str()).unwrap_or("tcp");
                let chain = args.get("chain").and_then(|v| v.as_str()).unwrap_or("input");
                let result = crate::command::run("sudo", &["nft", "add", "rule", "inet", "filter", chain, &format!("{}", proto), "dport", &port.to_string(), "accept"]).await?;
                Ok(serde_json::json!({ "action": "add_port", "port": port, "proto": proto, "chain": chain, "exit_code": result.exit_code }))
            }
            "remove_port" => {
                let port = args.get("port").and_then(|v| v.as_u64()).ok_or_else(|| ToolError::invalid_argument("Missing: port"))?;
                let proto = args.get("proto").and_then(|v| v.as_str()).unwrap_or("tcp");
                let chain = args.get("chain").and_then(|v| v.as_str()).unwrap_or("input");
                let handle_result = crate::command::run("sudo", &["nft", "--handle", "list", "chain", "inet", "filter", chain]).await?;
                let handle_re = regex::Regex::new(&format!(r".* {} dport {} accept # handle (\d+)", regex::escape(proto), port)).ok();
                let handle = handle_result.stdout.lines().find_map(|l| {
                    handle_re.as_ref().and_then(|re| re.captures(l)).and_then(|c| c.get(1)).map(|m| m.as_str().to_string())
                }).ok_or_else(|| ToolError::invalid_argument("Rule not found"))?;
                let result = crate::command::run("sudo", &["nft", "delete", "rule", "inet", "filter", chain, "handle", &handle]).await?;
                Ok(serde_json::json!({ "action": "remove_port", "port": port, "exit_code": result.exit_code }))
            }
            "validate" => {
                let result = crate::command::run("sudo", &["nft", "-c", "list", "ruleset"]).await?;
                Ok(serde_json::json!({ "valid": result.exit_code == 0, "output": result.stderr }))
            }
            "reload" => {
                let result = crate::command::run("sudo", &["systemctl", "reload", "nftables"]).await?;
                Ok(serde_json::json!({ "action": "reload", "exit_code": result.exit_code, "output": result.stdout }))
            }
            _ => Err(ToolError::invalid_argument("Unknown action")),
        }
    }
}
